File: 437-path-sum-iii/path-sum-iii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> int:
        
        def dfs(node, s):
            global t
            if node is None:
                return 0
            if s == node.val:
                t+=1
            dfs(node.left, s-node.val)
            dfs(node.right, s-node.val)
        
        def traverse(node, s):
            if node is None:
                return
            dfs(node, s)
            traverse(node.left, s)
            traverse(node.right, s)
        global t
        t = 0
        traverse(root, targetSum)
        return t
        

File: 437-path-sum-iii/path-sum-iii.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int count;
    int target;
    Map<Long, Integer> map;
    public int pathSum(TreeNode root, int targetSum) {
        target = targetSum;
        count = 0;
        map = new HashMap<>();
        dfs(root, 0L);
        return count;
    }

    public void dfs(TreeNode node, long curSum) {
        if(node == null) {
            return;
        }
        curSum += node.val;
        if(curSum == target) {
            count++;
        }
        count += map.getOrDefault(curSum - target, 0);
        map.put(curSum, map.getOrDefault(curSum, 0)+1);
        dfs(node.left, curSum);
        dfs(node.right, curSum);
        map.put(curSum, map.get(curSum)-1);
    }
}